app/controllers/courses_controller.rb

```diff
@@ -2,14 +2,73 @@ class CoursesController < ApplicationController
   before_action :authenticate_user!
   before_action :set_course, only: %i[ show edit update destroy ]
 
+
   # GET /courses or /courses.json
   def index
     @courses = Course.where(teacher: current_user.email)
   end
 
   # GET /courses/1 or /courses/1.json
   def show
-    @student_records = Student.where(course_id: params[:id])
+    #get the course id's for every past and present section of this course
+    @all_course_ids = Array[]
+    Course.where(course_name: Course.where(id: params[:id])[0].course_name).each do |c|
+      @all_course_ids.append(c.id)
+    end
+    #get all students currently and previously enrolled in this course
+    @student_records = Student.where(course_id: @all_course_ids, teacher: current_user.email)
+    #get all students tags for those currently and previously enrolled in this course
+    @tags = Set[]
+    for student in @student_records do
+        @tags.add(student.tags)
+    end
+    #get all the current and previous semesters and sections of this course
+    @semesters = Set[]
+    @sections = Set[]
+    for record in Course.all do
+        if record.course_name == @course.course_name
+            @semesters.add(record.semester)
+            @sections.add(record.section)
+        end
+    end
+    #if the user selects any dropdown menu filters
+    if params[:selected_semester].nil? == false or params[:selected_section].nil? == false or params[:selected_tag].nil? == false
+        #dropdown menu selections
+        @selected_tag = params[:selected_tag]
+        @selected_semester = params[:selected_semester]
+        @selected_section = params[:selected_section]
+        #get all course id's for the selected semester+section combo
+        if @selected_section == ''
+            @target_course_id = Course.where(course_name: Course.where(id: params[:id])[0].course_name,semester: @selected_semester)
+        else
+            @target_course_id = Course.where(course_name: Course.where(id: params[:id])[0].course_name,semester: @selected_semester, section: @selected_section)
+            #nil check in case the selected semester doesn't have the selected section
+            if @target_course_id.length > 0
+                @target_course_id = @target_course_id[0].id
+            end
+        end
+        #create the filtered list of students to display
+        if @selected_semester == '' and @selected_tag == ''
```

Comment:
is it requirement to select semester before section?

Reply:
As of right now no, because making that requirement would require the semester dropdown to have its own submit button. Otherwise the view won't know if the user selected a semester yet or not.